File: 15_caterpillar_method/min_abs_sum_of_two.cpp
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> &a) {
  int n = a.size();
  int rt = (int) 2e9;
  int beg = 0;
  int end = n - 1;
  sort(a.begin(), a.end());
  while (beg < end) {
    rt = min(rt, abs(a[beg] + a[end]));
    if (-a[beg] == a[end]) {
      return 0;
    } else if (abs(a[beg]) < abs(a[end])) {
      end--;
    } else {
      beg++;
    }
  }
  return min(rt, abs(a[beg] + a[end]));
}

File: 15_caterpillar_method/min_abs_sum_of_two.py
def solution(a):
  n = len(a)
  rt = int(2e9)
  beg = 0
  end = n - 1
  a.sort()
  while (beg < end):
    rt = min(rt, abs(a[beg] + a[end]))
    if -a[beg] == a[end]:
      return 0
    elif abs(a[beg]) < abs(a[end]):
      end -= 1
    else:
      beg += 1
  return min(rt, abs(a[beg] + a[end]))
